packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
+    notifyListeners();
+  }
+
+  Color? get backgroundCursorColor => _backgroundCursorColor;
+  Color? _backgroundCursorColor;
+  set backgroundCursorColor(Color? value) {
+    if (backgroundCursorColor?.value == value?.value)
+      return;
+
+    _backgroundCursorColor = value;
+    if (showRegularCaret)
+      notifyListeners();
+  }
+
+  Rect? get floatingCursorRect => _floatingCursorRect;
+  Rect? _floatingCursorRect;
+  set floatingCursorRect(Rect? value) {
+    if (_floatingCursorRect == value)
+      return;
+    _floatingCursorRect = value;
+    notifyListeners();
+  }
+
+  void paintRegularCursor(Canvas canvas, RenderEditable renderEditable, Color caretColor, TextPosition textPosition) {
+    final Rect caretPrototype = renderEditable._caretPrototype;
+    final Offset caretOffset = renderEditable._textPainter.getOffsetForCaret(textPosition, caretPrototype);
+    Rect caretRect = caretPrototype.shift(caretOffset + cursorOffset);
+
+    final double? caretHeight = renderEditable._textPainter.getFullHeightForCaret(textPosition, caretPrototype);
+    if (caretHeight != null) {
+      switch (defaultTargetPlatform) {
+        case TargetPlatform.iOS:
+        case TargetPlatform.macOS:
+          final double heightDiff = caretHeight - caretRect.height;
+          // Center the caret vertically along the text.
+          caretRect = Rect.fromLTWH(
+            caretRect.left,
+            caretRect.top + heightDiff / 2,
+            caretRect.width,
+            caretRect.height,
+          );
+          break;
+        case TargetPlatform.android:
+        case TargetPlatform.fuchsia:
+        case TargetPlatform.linux:
+        case TargetPlatform.windows:
+          // Override the height to take the full height of the glyph at the TextPosition
+          // when not on iOS. iOS has special handling that creates a taller caret.
+          // TODO(garyq): See the TODO for _computeCaretPrototype().
+          caretRect = Rect.fromLTWH(
+            caretRect.left,
+            caretRect.top - _kCaretHeightOffset,
+            caretRect.width,
+            caretHeight,
+          );
+          break;
+      }
+    }
+
+    caretRect = caretRect.shift(renderEditable._paintOffset);
+    final Rect integralRect = caretRect.shift(renderEditable._snapToPhysicalPixel(caretRect.topLeft));
+    final Radius? radius = cursorRadius;
+    caretPaint.color = caretColor;
+    if (radius == null) {
+      canvas.drawRect(integralRect, caretPaint);
+    } else {
+      final RRect caretRRect = RRect.fromRectAndRadius(integralRect, radius);
+      canvas.drawRRect(caretRRect, caretPaint);
+    }
+    caretPaintCallback(integralRect);
+  }
+
+  @override
+  void paint(Canvas canvas, Size size, RenderEditable renderEditable) {
+    if (!shouldPaint)
+      return;
+
+    assert(renderEditable != null);
+    final TextSelection? selection = renderEditable.selection;
+
+    if (selection == null || !selection.isCollapsed /* || !selection.isValid*/)
```

Comment:
I thought we shouldn't paint the caret when selection is (-1, -1)?

Reply:
I remember this came up awhile back in https://github.com/flutter/engine/pull/19785, was there somewhere else that we were talking about this?

Reply:
Not that I remember. Is the expected behavior of `(-1, -1)` documented anywhere? I couldn't seem to find anything about it.

The `TextEditingValue` constructor uses `(-1, -1)` as the default selection range for years. https://github.com/flutter/flutter/blame/d35a9db6fe439b1efaddffb19cea540b99917d16/packages/flutter/lib/src/services/text_input.dart#L82-L90

Reply:
No, I haven't seen any good docs about it.